server/src/instant/util/cache.clj

```diff
@@ -0,0 +1,106 @@
+(ns instant.util.cache
+  (:require [clojure.data.priority-map :as priority-map]
+            [clojure.core.cache :as cache :refer [defcache CacheProtocol]]
+            [instant.util.coll :refer [disj-in]]))
+
+(defn- build-leastness-queue
+  [base start-at]
+  (into (priority-map/priority-map) (for [[k _] base] [k start-at])))
+
+(defprotocol MultiEvictLRUCacheImpl
+  (evict-impl [this key]))
+
+;; mapping-fn takes an result in the cache and returns a new cache key
+;; n.b. the mapping-fn should return a key that doesn't overlap with a
+;; key that could potentially be in the cache.
+;; mapping is a hash map of {(mapping-fn cache-item) -> #{cache-keys}}
+(defcache MultiEvictLRUCache [cache lru mapping mapping-fn tick limit]
+  CacheProtocol
+  (lookup [_ item]
+    (get cache item))
+  (lookup [_ item not-found]
+    (get cache item not-found))
+  (has? [_ item]
+    (contains? cache item))
+  (hit [_ item]
+    (let [tick+ (inc tick)]
+      (MultiEvictLRUCache. cache
+                           (if (contains? cache item)
+                             (assoc lru item tick+)
+                             lru)
+                           mapping
+                           mapping-fn
+                           tick+
+                           limit)))
+  (miss [_ item result]
+    (let [tick+ (inc tick)]
+      (if (>= (count lru) limit)
+        (let [k (if (contains? lru item)
```

Comment:
@dwwoelfel curious, for the miss case why do we check whether the item is in the cache? Could we just make` k (first (peek lru))`?